### DISCUSSION TITLE
How does dockerized Directus handle migrations?

### DISCUSSION THREAD

**Discussion by k0l0ssus** [directus/directus#7393] [Archive]
Hey gang: I'm evaluating directus for a Dockerized deployment. I'm trying to understand what happens with the environment variables between migrations.
**Scenario:**

- I've deployed a brand-new installation of this fine product using the docker deployment model
   - Directus installs its own tables and I create a bunch of collections
   - I back up the database (along with the `directus_*` tables) and restore into a different database installation
- I run a new container instance against the new database installation
   - I define almost brand new everything for the config: admin credentials, redis server, S3 locations the works

Does the new Directus  install (against a backed up/migrated database) pick up my new config or does it retain the configs from the backed up DB or does it take note of the newer container environment vars? I ask because while I've looked at the `directus_*` tables, I don't see anything (in the settings table for example) that would indicate that the previous deployment's data was carried over. Yet, things like the admin credentials don't seem to have been picked up from the new Docker deployment. Thanks for your help.

**Comment by dstoyanoff** [comment]
Not a maintainer, but we work with Directus for quite some time now. 

Usually what you want to do is possible, but things like the admin credentials have some constraints. On the first scaffold of the application, the admin will be created and persisted in the database with it's password hashed.

When you create a second instance, Directus will see that your database is up to date and won't even try to execute the admin creation.

From what I've discovered so far, there are 2 options:
- manage all migrations manually - instead of doing things from the directus interface, do those with custom migrations. This is very scalable, but requires a lot of work
- have a generic admin user - basically have the same credentials of the user, but when you create the 2nd instance, go to the interface and create another admin

**Reply by rijkvanzanten** [reply]
This is correct. The docker image that we provide (`directus/directus`) will use `npx directus bootstrap` on first boot to make sure the database is installed and/or up to date.

> have a generic admin user - basically have the same credentials of the user, but when you create the 2nd instance, go to the interface and create another admin

You can use the `ADMIN_EMAIL` and `ADMIN_PASSWORD` environment variables for this. That'll set the email/password for the first user that is provisioned. By setting this to a known value, you make sure you never end up in a situation where Directus auto-generates the first user, and you lose the credentials it generates (it'll print it to stdout that first generation)

**Reply by dstoyanoff** [reply]
The thing about those env vars is that this is going to work on the first run on top of a database. If then this DB is backed up and restored to another server, even if you specify different email/password, those will be ignored and the original ones will be used. Correct me if I am wrong about it..

**Reply by rijkvanzanten** [reply]
> If then this DB is backed up and restored to another server, even if you specify different email/password, those will be ignored and the original ones will be used. Correct me if I am wrong about it..

Correct, if the database already contains users, no new admin user will be added 👍🏻

**Reply by k0l0ssus** [reply]
Thank you! Now, how about the `KEY` and `SECRET` vars? I'm not really clear on what they do, but I imagine they're part of a PKI setup? Can those two be safely modified between migrations?

**Reply by rijkvanzanten** [reply]
They're used to generate the JWT tokens. You'd want to have those unique per directus instance (they're not tied to the DB in any way). Just keep in mind that all users will be logged out once `SECRET` changes 👍🏻

**Reply by k0l0ssus** [reply]
Brilliant! If I'm reading the whole thing correctly, just the admin is necessarily db-bound; everything else can be modified without damaging an installation. Thanks @rijkvanzanten!
